src/node_crypto.cc

```diff
@@ -210,6 +210,43 @@ static int CryptoPemCallback(char *buf, int size, int rwflag, void *u) {
   return 0;
 }
 
+// Loads OpenSSL engine by engine id and returns it. The loaded engine
+// gets a reference so remember the corresponding call to ENGINE_free.
+// In case of error the appropriate js exception is scheduled
+// and nullptr is returned.
+#ifndef OPENSSL_NO_ENGINE
+static ENGINE* LoadEngineById(Environment* env, const char* engine_id) {
+  ENGINE* engine = ENGINE_by_id(engine_id);
+
+  MarkPopErrorOnReturn mark_pop_error_on_return;
+
+  if (engine == nullptr) {
+    // Engine not found, try loading dynamically.
+    engine = ENGINE_by_id("dynamic");
+    if (engine != nullptr) {
+      if (!ENGINE_ctrl_cmd_string(engine, "SO_PATH", engine_id, 0) ||
+          !ENGINE_ctrl_cmd_string(engine, "LOAD", nullptr, 0)) {
+        ENGINE_free(engine);
+        engine = nullptr;
+      }
+    }
+  }
+
+  if (engine == nullptr) {
+    char errmsg[1024];
+    int err = ERR_get_error();
+    if (err != 0) {
+      ERR_error_string_n(err, errmsg, sizeof(errmsg));
+    } else {
+      snprintf(errmsg, sizeof(errmsg),
+               "Engine \"%s\" was not found", engine_id);
+    }
+    env->ThrowError(errmsg);
```

Comment:
I still think moving the ThrowError out of this function is aesthetically nicer but I'll let it rest.